### DISCUSSION TITLE
批量插入若干批次后会报Deadline Exceeded

### DISCUSSION THREAD

**Discussion by levylll** [milvus-io/milvus#27939] [Q&A and General discussion]
插入的数据结构是dataframe  collection.insert(tmp_df, timeout=10)
若干批次后，会报错，报错如下。
`  File "/root/anaconda/lib/python3.10/site-packages/grpc/_channel.py", line 881, in result
    raise self
grpc._channel._MultiThreadedRendezvous: <_MultiThreadedRendezvous of RPC that terminated with:
	status = StatusCode.DEADLINE_EXCEEDED
	details = "Deadline Exceeded"
	debug_error_string = "UNKNOWN:Deadline Exceeded {grpc_status:4, created_time:"2023-10-26T03:59:02.914968372+00:00"}"
`
比较难受的是，milvus的线上服务也会挂掉重启，服务器的报错日志如下
`
goroutine 24541 [chan receive, locked to thread]:
runtime.gopark(0x0?, 0xc00456a701?, 0x0?, 0x0?, 0x357c2f6?)
	/usr/local/go/src/runtime/proc.go:381 +0xd6 fp=0xc00456a6d0 sp=0xc00456a6b0 pc=0x1543096
runtime.chanrecv(0xc006ee37a0, 0xc00456a7a8, 0x1)
	/usr/local/go/src/runtime/chan.go:583 +0x49d fp=0xc00456a760 sp=0xc00456a6d0 pc=0x150b57d
runtime.chanrecv2(0xc005ba4aa0?, 0xc007ade5d0?)
	/usr/local/go/src/runtime/chan.go:447 +0x18 fp=0xc00456a788 sp=0xc00456a760 pc=0x150b0b8
github.com/panjf2000/ants/v2.(*goWorker).run.func1()
	/go/pkg/mod/github.com/panjf2000/ants/v2@v2.7.2/worker.go:63 +0x7b fp=0xc00456a7e0 sp=0xc00456a788 pc=0x28cb75b
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1598 +0x1 fp=0xc00456a7e8 sp=0xc00456a7e0 pc=0x1579301
created by github.com/panjf2000/ants/v2.(*goWorker).run
	/go/pkg/mod/github.com/panjf2000/ants/v2@v2.7.2/worker.go:48 +0x65
`
请问这种情况是为什么呢？尝试调整了批次和每个批次间的sleep，都解决不了问题，一段时间后还是会挂掉。

**Comment by yhmo** [comment]
It could be OOM. But the logs you show didn't reveal the root cause.
Some questions:
- standalone or cluster? how much memory is assigned to each node(querynode/datanode)?
- how many collections in the server? how many data rows were loaded into memory?
- the last minute log of the server

**Reply by levylll** [reply]
昨天加了内存，之前是8G现在加到了32G，之前一直没法灌数据成功的问题似乎解决了。
但是现在有个问题，查询时会timeout，因为我每次灌数据之前会批量根据某个字段查询一次（针对非主键的查询），如果已经写入就不再写入。
查询方法如下：
res = collection.query(expr="o_id in %s" % o_ids, timeout=10, output_fields=["o_id"])

报错如下：
debug_error_string = "UNKNOWN:failed to connect to all addresses; last error: UNKNOWN: ipv4:*:*: Failed to connect to remote host: Connection refused {created_time:"2023-10-27T02:56:25.516691149+00:00", grpc_status:14}"

是不是这种查询对milvus压力太大了，有没有更好的解决方案呢？

**Reply by levylll** [reply]
在日志中找到了报错，麻烦帮忙看一下呢


panic: runtime error: makeslice: len out of range

goroutine 226097 [running]:
panic({0x3b67ca0, 0x4490880})
	/usr/local/go/src/runtime/panic.go:987 +0x3bb fp=0xc000a22280 sp=0xc000a221c0 pc=0x153fbdb
runtime.panicmakeslicelen(...)
	/usr/local/go/src/runtime/slice.go:29
runtime.makeslice(0xc000a222e8?, 0x26e589e?, 0xc000a222d8?)
	/usr/local/go/src/runtime/slice.go:98 +0x78 fp=0xc000a222a8 sp=0xc000a22280 pc=0x1558cb8
[github.com/bits-and-blooms/bitset.(*BitSet).extendSetMaybe](http://github.com/bits-and-blooms/bitset.(*BitSet).extendSetMaybe)(0xc0018ef940, 0x34479f794177ab60?)
	/go/pkg/mod/[github.com/bits-and-blooms/bitset@v1.2.0/bitset.go:150](http://github.com/bits-and-blooms/bitset@v1.2.0/bitset.go:150) +0xa9 fp=0xc000a222f8 sp=0xc000a222a8 pc=0x26e2289
[github.com/bits-and-blooms/bitset.(*BitSet).Set(...)](http://github.com/bits-and-blooms/bitset.(*BitSet).Set(...))
	/go/pkg/mod/[github.com/bits-and-blooms/bitset@v1.2.0/bitset.go:173](http://github.com/bits-and-blooms/bitset@v1.2.0/bitset.go:173)
[github.com/bits-and-blooms/bloom/v3.(*BloomFilter).Add](http://github.com/bits-and-blooms/bloom/v3.(*BloomFilter).Add)(0xc0035595d8, {0xc000a223c0?, 0xc009662bf8?, 0xffffffffffffffff?})
	/go/pkg/mod/[github.com/bits-and-blooms/bloom/v3@v3.0.1/bloom.go:142](http://github.com/bits-and-blooms/bloom/v3@v3.0.1/bloom.go:142) +0xf7 fp=0xc000a22398 sp=0xc000a222f8 pc=0x26e5d57
[github.com/milvus-io/milvus/internal/storage.(*PrimaryKeyStats).Update](http://github.com/milvus-io/milvus/internal/storage.(*PrimaryKeyStats).Update)(0xc002052550, {0x44cb610, 0xc009662bf8})
	/go/src/[github.com/milvus-io/milvus/internal/storage/stats.go:168](http://github.com/milvus-io/milvus/internal/storage/stats.go:168) +0xfd fp=0xc000a22400 sp=0xc000a22398 pc=0x2cbc79d
[github.com/milvus-io/milvus/internal/datanode.(*compactionTask).merge](http://github.com/milvus-io/milvus/internal/datanode.(*compactionTask).merge)(0xc002e67e40, {0x44b3d08, 0xc0035afb90}, {0xc003ef2900, 0x8, 0x44c20a0?}, 0xc002e67e40?, 0xc00254f3e8?, 0xc0032f2800, 0xc002cfc8a0)
	/go/src/[github.com/milvus-io/milvus/internal/datanode/compactor.go:455](http://github.com/milvus-io/milvus/internal/datanode/compactor.go:455) +0x1fa8 fp=0xc000a23158 sp=0xc000a22400 pc=0x30659a8
[github.com/milvus-io/milvus/internal/datanode.(*compactionTask).compact(0xc002e67e40)](http://github.com/milvus-io/milvus/internal/datanode.(*compactionTask).compact(0xc002e67e40))
	/go/src/[github.com/milvus-io/milvus/internal/datanode/compactor.go:733](http://github.com/milvus-io/milvus/internal/datanode/compactor.go:733) +0x1287 fp=0xc000a23d60 sp=0xc000a23158 pc=0x3067747
[github.com/milvus-io/milvus/internal/datanode.(*compactionExecutor).executeTask](http://github.com/milvus-io/milvus/internal/datanode.(*compactionExecutor).executeTask)(0xc001134270, {0x44c20a0, 0xc002e67e40})
	/go/src/[github.com/milvus-io/milvus/internal/datanode/compaction_executor.go:96](http://github.com/milvus-io/milvus/internal/datanode/compaction_executor.go:96) +0x315 fp=0xc000a23fb8 sp=0xc000a23d60 pc=0x3061675
[github.com/milvus-io/milvus/internal/datanode.(*compactionExecutor).executeWithState.func1()](http://github.com/milvus-io/milvus/internal/datanode.(*compactionExecutor).executeWithState.func1())
	/go/src/[github.com/milvus-io/milvus/internal/datanode/compaction_executor.go:75](http://github.com/milvus-io/milvus/internal/datanode/compaction_executor.go:75) +0x2e fp=0xc000a23fe0 sp=0xc000a23fb8 pc=0x306122e
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1598 +0x1 fp=0xc000a23fe8 sp=0xc000a23fe0 pc=0x1579301
created by [github.com/milvus-io/milvus/internal/datanode.(*compactionExecutor).executeWithState](http://github.com/milvus-io/milvus/internal/datanode.(*compactionExecutor).executeWithState)
	/go/src/[github.com/milvus-io/milvus/internal/datanode/compaction_executor.go:75](http://github.com/milvus-io/milvus/internal/datanode/compaction_executor.go:75) +0x95

**Reply by yhmo** [reply]
makeslice: len out of range

像是触发了代码里某个bug，得往前翻一翻日志。

**expr="o_id in %s" % o_ids**
你确定你的expr是这么输入的？

**Reply by levylll** [reply]
对啊，直接复制的代码（变量名缩写了一下）。是我这样写是不合理的吗？
并且如果查询循环次数少，不会出问题，循环的次数多了之后才会出这个问题。
那我们再翻一下日志看看。

**Reply by yhmo** [reply]
@levylll 
已经有好心人帮忙修复这个问题了。我感觉这个问题在所有2.2/2.3版本中都存在，并且是个特定条件下才会触发的bug。
我想了解一下你的具体操作是怎样的，以便让我们的测试人员尝试复现。

**Reply by levylll** [reply]
为了防止数据插重复，每次插入之前，都会批量查询这批id是否已经存在（id非主键，非唯一），如果不存在则插入，如果存在则跳过，插入时，这个collection中大概有几万条数据了。
每个批次512条，每个批次间隔1s
id的定义`
FieldSchema(
            name="o_id",
            dtype=DataType.INT64,
            is_primary=True,
        )`

**Comment by ulovecode** [comment]
github.com/bits-and-blooms/bitset 这个版本的库有问题，会导致分配直接超过golang的切片长度限制。使用最新版本的 github.com/bits-and-blooms/bitset  库没有出现这个问题。帮忙合并一下 https://github.com/milvus-io/milvus/pull/27984

newset := make([]uint64, nsize, 2*nsize) // increase capacity 2x


func (b *BitSet) extendSet(i uint) {
	if i >= Cap() {
		panic("You are exceeding the capacity")
	}
	nsize := wordsNeeded(i + 1)
	if b.set == nil {
		b.set = make([]uint64, nsize)
	} else if cap(b.set) >= nsize {
		b.set = b.set[:nsize] // fast resize
	} else if len(b.set) < nsize {
		newset := make([]uint64, nsize, 2*nsize) // increase capacity 2x
		copy(newset, b.set)
		b.set = newset
	}
	b.length = i + 1
}

**Reply by yhmo** [reply]
请问bitset这个bug一般是什么情况下触发？ bloom.Add()获得的输入是一个由int64的值，在内部经过hash以及其他转化后得到一个uint，然后进入bitset.extendSet，是不是因为newset := make([]uint64, nsize, 2nsize) 这行有可能会出这个panic？

**Reply by ulovecode** [reply]
对的

**Comment by levylll** [comment]
更新了最新的版本后，这个错误还是会有。有没有办法我们自己起来milvus服务后，可以自己debug milvus代码的方式呢？

**Reply by yhmo** [reply]
你是用v2.3.3还是不行么？报错不是之前那个错吧？

**Reply by yanliang567** [reply]
我们正在尝试复现这个问题，请问一下，你说的若干次一般是多少次呢，插入数据的主键是否重复

**Reply by levylll** [reply]
一两百个批次后会报错，每个批次会针对512条数据进行一次查询（非主键查询），插入不存在的条目，插入的主键 定义如下 `        
id = FieldSchema(
            name="id",
            dtype=DataType.INT64,
            is_primary=True,
            auto_id=True,
        )`

**Reply by yhmo** [reply]
问题仍然存在的原因是因为[那个pr](https://github.com/milvus-io/milvus/pull/27984)没有合进2.3/2.2分支 ，所以最新的版本里没有这个fix
今天刚合进2.3/2.2分支，后续版本v2.3.4/v2.2.15才会有这个fix
https://github.com/milvus-io/milvus/pull/28412
https://github.com/milvus-io/milvus/pull/28411

**Comment by xiaofan-luan** [comment]
@yhmo 
I think we should add a lock in PKfilter Test and PKFilter.AddString
This might not be thread safe

**Reply by xiaofan-luan** [reply]
seems that this pr is not merged to 2.3.3

**Reply by levylll** [reply]
请问2.3.3什么时候发布。现在线上用的有点难受了，要么数据灌不进去，要么脚本显示已经灌进去了，实际上milvus已经崩了，然后其实灌失败了，后面查询发现有些数据根本没灌成功，导致线上也不确定哪些灌成功哪些没灌成功。 @xiaofan-luan

**Reply by yhmo** [reply]
@levylll 
估计2.3.4还有一两周才能发。你有没有具体的重现脚本？我们这边尝试了没能重现你说的问题。而且看起来只有你遇到了，其他用户没遇到。

**Reply by whNLP** [reply]
我也遇到这种情况

**Comment by levylll** [comment]
> @levylll 估计2.3.4还有一两周才能发。你有没有具体的重现脚本？我们这边尝试了没能重现你说的问题。而且看起来只有你遇到了，其他用户没遇到。
这个比较奇怪的是，又不一定说能稳定复现，我们有尝试过本地重新搞一个milvus，批量插入数据也没有问题，但是线上实际使用时，就插入几万条后就线上milvus就挂了。。。有点怀疑是不是我们之前的某些数据可能影响到了milvus的稳定性。

**Reply by yhmo** [reply]
每次挂的时候，日志里都是你说的这样么？ https://github.com/milvus-io/milvus/discussions/27939#discussioncomment-7399062

**Reply by levylll** [reply]
对的

**Reply by xiaofan-luan** [reply]
我们下周一发2.3.4会修复这个问题

**Reply by levylll** [reply]
两周过去了，请问什么时候更新啊，现在想新创建个collection已经不行了，灌了5000条就崩了。。。很难受啊。

**Reply by levylll** [reply]
@xiaofan-luan xiao

**Reply by yhmo** [reply]
2.3.4还有些问题没解决，还得再等等。我还是很好奇为何只有你经常遇到这个问题，有没有试过改变一下导入数据的行为。比如不检查，直接insert。

**Comment by levylll** [comment]
> 2.3.4还有些问题没解决，还得再等等。我还是很好奇为何只有你经常遇到这个问题，有没有试过改变一下导入数据的行为。比如不检查，直接insert。
嗯，现在就是直接insert，没有检查了，但是还是会报 @yhmo

**Reply by yhmo** [reply]
你可以先在测试环境里用这个镜像起milvus试验一下。这是我们为2.3.4测试用的镜像，包含那个修复。
`docker pull milvusdb/milvus:2.3-20231220-f1ed04be-amd64`

然后pymilvus用2.3.3版本。看是否还有问题。

**Reply by levylll** [reply]
好的，我们这边看看能不能试一下。非常感谢！
